docs/docs/propeller/streams-sink.md

```diff
@@ -0,0 +1,37 @@
+# StreamsSink Architecture
+
+## AsyncQueue
+
+The StreamsSink uses an AsyncQueue that provides a node-friendly way to handle a
+queue of items without blocking waits.
+
+It is backed by a basic FIFO Queue, implemented using a linked list. When an
+item is added to the AsyncQueue, it first checks if there are any pending
+retrieval requests. If there are, it fulfills the oldest request. If not, it
+queues the item. The dequeueing operation retrieves an item from the queue. If
+the queue is empty, it waits asynchronously until an item is available or until
+an abort signal is triggered, which would stop the wait and reject the promise
+
+## Receiving batches
+
+For every event within a batch, the StreamsSink checks if its associated Stream
+is already in the processing queue. If it is, the event is combined with it,
+ensuring it's processed alongside previously queued events.
+
+The set of streams associated with the batch are stored in a `Set` so we can
+know when a batch has finished processing and notify the submitter.
+
+The StreamsSink limits the number of in-flight batches to a configured number.
+When a batch is submitted we check the number of in-flight batches and if it has
+reached the threshold delays the introduction of the batch's events until a
```

Comment:
>delays the introduction of the batch's events
wrong semantic - if the work has been done to read, it should be fed into the buffer immediately as the memory is being consumed
it needs to wait _after_ ingesting the batch, allowing completion of a batch to open up capacity to read mode